node/src/store.rs

```diff
@@ -135,4 +174,144 @@ pub enum StoreError {
 
     #[error("Error opening store: {0}")]
     OpenFailed(String),
+
+    #[error("Invalid headers range")]
+    InvalidHeadersRange,
+}
+
+/// a helper function to convert any kind of range to the inclusive range of header heights.
+fn to_headers_range(bounds: impl RangeBounds<u64>, last_index: u64) -> Result<RangeInclusive<u64>> {
+    let start = match bounds.start_bound() {
+        // in case of unbounded, default to the first height
+        Bound::Unbounded => 1,
+        // range starts after the last index or before first height
+        Bound::Included(&x) if x > last_index || x == 0 => return Err(StoreError::NotFound),
+        Bound::Excluded(&x) if x >= last_index => return Err(StoreError::NotFound),
+        // valid start indexes
+        Bound::Included(&x) => x,
+        Bound::Excluded(&x) => x + 1, // can't overflow thanks to last_index check
+    };
+    let end = match bounds.end_bound() {
+        // in case of unbounded, default to the last index
+        Bound::Unbounded => last_index,
+        // range ends after the last index
+        Bound::Included(&x) if x > last_index => return Err(StoreError::NotFound),
+        Bound::Excluded(&x) if x > last_index + 1 => return Err(StoreError::NotFound),
```

Comment:
```suggestion
        Bound::Excluded(&x) if x >= last_index => return Err(StoreError::NotFound),
```

Reply:
this is not equivalent :) in fact this re-introduces the bug I had before

Reply:
Excluded(last_index + 1) == last_index, the logic here is correct

Reply:
actually this is wrong. i dont know what i was thinking